datafusion/functions-nested/src/string.rs

```diff
@@ -495,20 +505,173 @@ pub(super) fn array_to_string_inner(args: &[ArrayRef]) -> Result<ArrayRef> {
+        }
+        LargeUtf8 => {
+            let string_array = args[0].as_string::<T>();
+            let builder = LargeStringBuilder::with_capacity(string_array.len(), string_array.get_buffer_memory_size());
+            string_to_array_inner_2::<&GenericStringArray<T>, LargeStringBuilder>(args, string_array, builder)
+        }
+        other =>  exec_err!("unsupported type for first argument to string_to_array function as {other:?}")
+    }
+}
+
+fn string_to_array_inner_2<'a, StringArrType, StringBuilderType>(
+    args: &'a [ArrayRef],
+    string_array: StringArrType,
+    string_builder: StringBuilderType,
+) -> Result<ArrayRef>
+where
+    StringArrType: StringArrayType<'a>,
+    StringBuilderType: StringArrayBuilderType,
+{
+    match args[1].data_type() {
+        Utf8 => {
+            let delimiter_array = args[1].as_string::<i32>();
+            if args.len() == 2 {
+                string_to_array_impl::<
+                    StringArrType,
+                    &GenericStringArray<i32>,
+                    &StringViewArray,
+                    StringBuilderType,
+                >(string_array, delimiter_array, None, string_builder)
+            } else {
+                string_to_array_inner_3::<StringArrType,
+                    &GenericStringArray<i32>,
+                    StringBuilderType>(args, string_array, delimiter_array, string_builder)
+            }
+        }
+        Utf8View => {
+            let delimiter_array = args[1].as_string_view();
+
+            if args.len() == 2 {
+                string_to_array_impl::<
+                    StringArrType,
+                    &StringViewArray,
+                    &StringViewArray,
+                    StringBuilderType,
+                >(string_array, delimiter_array, None, string_builder)
+            } else {
+                string_to_array_inner_3::<StringArrType,
+                    &StringViewArray,
+                    StringBuilderType>(args, string_array, delimiter_array, string_builder)
+            }
+        }
+        LargeUtf8 => {
+            let delimiter_array = args[1].as_string::<i64>();
+            if args.len() == 2 {
+                string_to_array_impl::<
+                    StringArrType,
+                    &GenericStringArray<i64>,
+                    &StringViewArray,
+                    StringBuilderType,
+                >(string_array, delimiter_array, None, string_builder)
+            } else {
+                string_to_array_inner_3::<StringArrType,
+                    &GenericStringArray<i64>,
+                    StringBuilderType>(args, string_array, delimiter_array, string_builder)
+            }
+        }
+        other =>  exec_err!("unsupported type for second argument to string_to_array function as {other:?}")
+    }
+}
 
-    match args.len() {
-        2 => {
+fn string_to_array_inner_3<'a, StringArrType, DelimiterArrType, StringBuilderType>(
```

Comment:
If I am reading this correctly, this function effectively handles all possible combinations of `Utf8`, `Utf8View` and `LargeUtf8` by generating specialized implementations for each combination -- so I think that means 3 * 3 = 9 copies of the function

I wonder if it is common to mix all the types 🤔